### PR TITLE
Fake: fix conv_transpose2d striding

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82846

The output striding channels-last preservation logic differs between cuda and cpu. For the meta kernel, we can peek at the fake tensor device and use that to determine whether to do cpu or cuda. 

You could argue there's a leaking of abstraction here but this seems like a pretty minimal leak and I'm not sure there's a much cleaner way forward for device-specific striding tracing logic.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
input sniffing is cool, we do it in other places too.

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -307,10 +307,17 @@ def calc_conv_nd_return_shape(
...
-    def pick_memory_format():
-        if input_tensor.is_contiguous(memory_format=torch.channels_last):
-            return torch.channels_last
-        elif input_tensor.is_contiguous(memory_format=torch.contiguous_format):
+    def is_channels_last(ten):
+        return ten.is_contiguous(memory_format=torch.channels_last)
+
+    def pick_memory_format(device_hint):
+        if device_hint == "cuda":
+            if is_channels_last(input_tensor) and is_channels_last(weight):
```
Comment (ngimel) [inline torch/_meta_registrations.py:315]: I see cuda producing channels-last if only weight is channels last
```
In [8]: m=m.cuda().to(memory_format=torch.channels_last)

In [9]: inp=inp.cuda()

In [10]: m(inp).stride()
Out[10]: (968, 1, 88, 8)

In [11]: inp.stride()
Out[11]: (144, 36, 6, 1)
```

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -307,10 +307,17 @@ def calc_conv_nd_return_shape(
                 )
         return ret_shape
 
-    def pick_memory_format():
-        if input_tensor.is_contiguous(memory_format=torch.channels_last):
-            return torch.channels_last
-        elif input_tensor.is_contiguous(memory_format=torch.contiguous_format):
+    def is_channels_last(ten):
+        return ten.is_contiguous(memory_format=torch.channels_last)
```
Comment (ngimel) [inline torch/_meta_registrations.py:311]: This is too permissive, e.g. weight tensors with 1x1 kernel will be channels-last contiguous, but they won't be channels-last for the purposes of memory format propagation

**Review by ngimel — COMMENTED** [review]
The right thing would be to expose `suggest_memory_format` to python because this is what's used internally to decide whether something is channels-last. But for now probably a TODO will suffice

File: torch/_meta_registrations.py (hunk 1 above)
Comment (eellison) [reply torch/_meta_registrations.py:315]: Yea, it's either, misread the `||` as an `&&`:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ConvUtils.h#L316

Comment (ngimel) [reply torch/_meta_registrations.py]: Which means that all 1x1 convs will happily return channels-last format :-(

Comment (eellison) [reply torch/_meta_registrations.py]: Which makes this PR correct at least :-)

**Comment by ezyang** [comment]
This might be the root cause of https://github.com/pytorch/torchdynamo/issues/1274 it would be good to land it

**Comment by eellison** [comment]
Yup, that's the intention. But also setting up cross-ref testing now..

**Pull request closed without merging**